src/Chainweb/Cut/Create.hs

```diff
@@ -0,0 +1,442 @@
+--
+-- 2. This is a transition step (parentGraph /= targetGraph). Only do the step
+--    when new /and/ old adjacents headers are available. The old dependencies
+--    are not used but are needed to guarantee that the cut can extended on
+--    other chains.
+--
+-- 3. This is a "normal" transition.
+--
+-- TODO: it is important that the semantics of this function corresponds to the
+-- respective validation in the module "Chainweb.Cut", in particular
+-- 'isMonotonicCutExtension'. It must not happen, that a cut passes validation
+-- that can't be further extended.
+--
+getCutExtension
+    :: HasCallStack
+    => HasChainId cid
+    => Cut
+        -- ^ the cut which is to be extended
+    -> cid
+        -- ^ the header onto which the new block is created. It is expected
+        -- that this header is contained in the cut.
+    -> Maybe CutExtension
+getCutExtension c cid = do
+
+    -- In a graph transition we wait for all chains to do the transition
+    -- to the new graph before moving ahead.
+    guard (not $ isGraphTransitionCut && isGraphTransitionPost)
+
+    -- If this is a graph transition step we have to check braiding for the
+    -- old /and/ the new graph.
+    when isGraphTransitionPre (void $ newAdjHashes parentGraph)
+
+    -- Finally, compute the adjacent hashes for the target graph.
+    as <- BlockHashRecord <$> newAdjHashes targetGraph
+
+    return CutExtension
+        { _cutExtensionCut' = c
+        , _cutExtensionParent' = ParentHeader p
+        , _cutExtensionAdjacentHashes' = as
+        }
+  where
+    p = c ^?! ixg (_chainId cid)
+    v = _chainwebVersion c
+    parentHeight = _blockHeight p
+    targetHeight = parentHeight + 1
+    parentGraph = chainGraphAt_ p parentHeight
+    targetGraph = chainGraphAt_ p targetHeight
+    isGraphTransitionPre = isGraphChange c targetHeight
+    isGraphTransitionPost = isGraphChange c parentHeight
+    isGraphTransitionCut = isTransitionCut c
+
+    -- | Try to get all adjacent hashes dependencies for the given graph.
+    --
+    newAdjHashes :: ChainGraph -> Maybe (HM.HashMap ChainId BlockHash)
+    newAdjHashes g =
+        imapM (\xcid _ -> hashForChain xcid)
+        $ HS.toMap
+        $ adjacentChainIds g p
+
+    hashForChain acid
+        -- existing chain
+        | Just b <- c ^? ixg acid = tryAdj b
+        -- new chain after graph transition
+        | targetHeight == genesisHeight v acid = Just $ genesisParentBlockHash v acid
+        | otherwise = error $ T.unpack
+            $ "getAdjacentParents: invalid cut, can't find adjacent hash for chain "
+            <> sshow acid
+            <> ". Parent: " <> encodeToText (ObjectEncoded p)
+
+    tryAdj :: BlockHeader -> Maybe BlockHash
+    tryAdj b
+
+        -- When the block is behind, we can move ahead
+        | _blockHeight b == targetHeight = Just $! _blockParent b
+
+        -- if the block is ahead it's blocked
+        | _blockHeight b + 1 == parentHeight = Nothing -- chain is blocked
```

Comment:
The comment above about returning an `Either XXX` would mean a lot of these comments could instead be code ie `Left "chain is blocked"`